### PR TITLE
[SYCL-MLIR] Allow inner type to be i8 pointer in `InitializeValueByInitListExpr`

### PR DESCRIPTION
**Pull request by whitneywhtsang**
This PR makes the following tests pass:
```
SYCL :: GroupAlgorithm/SYCL2020/exclusive_scan.cpp
SYCL :: GroupAlgorithm/SYCL2020/inclusive_scan.cpp
SYCL :: GroupAlgorithm/SYCL2020/reduce.cpp
SYCL :: GroupAlgorithm/exclusive_scan.cpp
SYCL :: GroupAlgorithm/inclusive_scan.cpp
SYCL :: GroupAlgorithm/reduce.cpp
SYCL :: SubGroup/reduce_fp16.cpp
SYCL :: SubGroup/reduce_fp64.cpp
SYCL :: SubGroup/scan_fp16.cpp
SYCL :: SubGroup/scan_fp64.cpp
SYCL :: SubGroup/scan_spirv13_fp64.cpp
```

Signed-off-by: Tsang, Whitney [whitney.tsang@intel.com](mailto:whitney.tsang@intel.com)

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 1]
```diff
@@ -290,21 +290,24 @@ mlir::Attribute MLIRScanner::InitializeValueByInitListExpr(mlir::Value ToInit,
...
+                Loc,
+                LLVM::LLVMPointerType::get(AT.getElementType(),
+                                           PT.getAddressSpace()),
+                ToInit, llvm::ArrayRef<mlir::LLVM::GEPArg>{0, I});
+          else if (ET.isInteger(8))
+            Next = Builder.create<LLVM::GEPOp>(
+                Loc, LLVM::LLVMPointerType::get(ET, PT.getAddressSpace()),
+                ToInit, llvm::ArrayRef<mlir::LLVM::GEPArg>{I});
           else
             llvm_unreachable("unknown inner type");
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:310]: NIT: Could refactor this using `TypeSwitch`

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 2]
```diff
@@ -290,21 +290,24 @@ mlir::Attribute MLIRScanner::InitializeValueByInitListExpr(mlir::Value ToInit,
...
-            NextType = AT.getElementType();
+            Next = Builder.create<LLVM::GEPOp>(
+                Loc,
+                LLVM::LLVMPointerType::get(AT.getElementType(),
+                                           PT.getAddressSpace()),
+                ToInit, llvm::ArrayRef<mlir::LLVM::GEPArg>{0, I});
+          else if (ET.isInteger(8))
+            Next = Builder.create<LLVM::GEPOp>(
+                Loc, LLVM::LLVMPointerType::get(ET, PT.getAddressSpace()),
+                ToInit, llvm::ArrayRef<mlir::LLVM::GEPArg>{I});
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:308]: Would this correspond to a `void *` pointer?

**Review by victor-eds — COMMENTED** [review]
Can we have a test for these changes? Also, I'd split in two PRs:
- `Maximum memref`
- Allow `i8 *`

**Comment by whitneywhtsang** [comment]
> Can we have a test for these changes? 

Was thinking to simply use the test suite tests as coverage. 

> Also, I'd split in two PRs:
> 
> * `Maximum memref`
> * Allow `i8 *`

Spitted the maximum memref change to https://github.com/intel/llvm/pull/8091.

**Comment by victor-eds** [comment]
> Was thinking to simply use the test suite tests as coverage.

I'd argue we should also have lit tests to check the actual code we're generating.

File: polygeist/tools/cgeist/Test/Verification/sycl/minimum.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
+// RUN: clang++ -fsycl -fsycl-device-only -O0 -w -emit-mlir %s -o - | FileCheck %s
+
+// CHECK-LABEL:    func.func @_ZNKSt4lessIvEclIRiS2_EEDTltclsr3stdE7forwardIT_Efp_Eclsr3stdE7forwardIT0_Efp0_EEOS3_OS4_(%arg0: !llvm.ptr<struct<(i8)>, 4> {llvm.align = 1 : i64, llvm.dereferenceable_or_null = 1 : i64, llvm.noundef}, %arg1: memref<?xi32, 4> {llvm.align = 4 : i64, llvm.dereferenceable = 4 : i64, llvm.noundef}, %arg2: memref<?xi32, 4> {llvm.align = 4 : i64, llvm.dereferenceable = 4 : i64, llvm.noundef}) -> (i1 {llvm.noundef, llvm.zeroext})
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Test/Verification/sycl/minimum.cpp:3]: The gep that generated got cleaned up:
```
          %11 = "llvm.getelementptr"(%1) {rawConstantIndices = array<i32: 0, 0>} : (!llvm.ptr<struct<(i8)>>) -> !llvm.ptr<i8>
          %12 = "llvm.getelementptr"(%11) {rawConstantIndices = array<i32: 0>} : (!llvm.ptr<i8>) -> !llvm.ptr<i8>
          %13 = "llvm.getelementptr"(%12) {rawConstantIndices = array<i32: 0>} : (!llvm.ptr<i8>) -> !llvm.ptr<i8>
```

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 2 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/CGExpr.cc]: Yes. `void *` is represented as `i8*`.

**Review by victor-eds — APPROVED** [review]

**Pull request merged**